### PR TITLE
feat: partly implemented rpc adapter

### PR DESCRIPTION
**Pull request by PeterBolha**
implemented several methods from perun rpc adapter

### CODE REVIEW

**Review by vyskocilpavel — CHANGES_REQUESTED** [review]
I think that the ApiClient object must be pass as the param to the `*ManagerApi` (https://github.com/CESNET/proxyaai_perun_connection_manager/blob/7d06cf12f14d5579d2675b12dfd10483b1416e89/perun_openapi/api_client.py#L43) with the specified configuration(Need to be added in some configuration file with items RPC hosts, username, password, etc.) 

But maybe I don't understand how to work with the generated RPC connectors.

File: Adapters/PerunRpcAdapter.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            if user["titleBefore"]:
+                name += user["titleBefore"] + " "
+            if user["firstName"]:
+                name += user["firstName"] + " "
+            if user["middleName"]:
+                name += user["middleName"] + " "
+            if user["lastName"]:
+                name += user["lastName"]
+            if user["titleAfter"]:
+                name += " " + user["titleAfter"]
```
Comment (melanger) [inline Adapters/PerunRpcAdapter.py:33]:
```suggestion
            for user_attr in ["titleBefore", "firstName", "middleName", "lastName", "titleAfter"]:
                name += user[user_attr] + " "
            name = name.strip()
```

File: Adapters/PerunRpcAdapter.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        resources = []
+        for perun_attr in perun_attrs:
+            resources.append(
+                Resource(
+                    perun_attr["id"],
+                    perun_attr["vo_id"],
+                    perun_attr["facility_id"],
+                    perun_attr["name"],
+                )
+            )
```
Comment (melanger) [inline Adapters/PerunRpcAdapter.py:79]:
```suggestion
        resources = [Resource(perun_attr["id"], perun_attr["vo_id"], perun_attr["facility_id"], perun_attr["name"]) for perun_attr in perun_attrs]
```
use list comprehension where possible

File: Adapters/PerunRpcAdapter.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+        return Vo(vo["id"], vo["name"], vo["short_name"])
+
+    def get_facility_by_rp_identifier(
+        self, rp_identifier: str, entity_id_attr="perunFacilityAttr_entityID"
+    ) -> Facility:
+        attr_name = AttributeUtils.get_rpc_attr_name(entity_id_attr)
+
+        if not attr_name:
```
Comment (melanger) [inline Adapters/PerunRpcAdapter.py:137]: If we want defaults, I think they should be handled by AttributeUtils, not here. That will reduce similar code quite significantly.

File: Adapters/PerunRpcAdapter.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                        group["description"],
+                    )
+                )
+
+        return self.remove_duplicate_entities(sp_groups)
+
+    def get_group_by_name(self, vo: Vo, name: str) -> Group:
+        group = GroupsManagerApi.get_group_by_name(self, vo, name)
+        attr = AttributesManagerApi.get_attribute(group["id"])
+        unique_name = attr["value"] + ":" + group["name"]
```
Comment (melanger) [inline Adapters/PerunRpcAdapter.py:104]: This is repeated quite a lot, please create a local function like `get_attribute_value`, which will get the attribute and return its `["value"]` directly.

File: Adapters/PerunRpcAdapter.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        return Group(
+            group["id"],
```
Comment (melanger) [inline Adapters/PerunRpcAdapter.py:106]: when the dictionary keys match the argument names of the constructor, you can use `Group(*group)` instead of writing individual items (this applies to all constructor calls)

**Review by melanger — CHANGES_REQUESTED** [review]

File: Adapters/PerunRpcAdapter.py [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+        #  instantiated here or passed to constructor.
+
+        self.logger = Logger.get_logger(self.__class__.__name__)
+
+        try:
+            self.CONFIG = Configuration(
+                config_data["rpc_url"],
+                config_data["rpc_username"],
+                config_data["rpc_password"],
+            )
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:44]:
```suggestion
            self.CONFIG = Configuration(
                host=config_data["rpc_url"],
                username=config_data["rpc_username"],
                password=config_data["rpc_password"],
            )
```

File: Adapters/PerunRpcAdapter.py [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+
+                return User(**user)
+
+    def get_group_unique_name(
+        self,
+        attributes_api_instance: AttributesManagerApi,
+        group_name: str,
+        group_id: int,
+    ):
+        attr = attributes_api_instance.get_attribute(group_id)
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:84]:
```suggestion
        attr = attributes_api_instance.get_attribute(
            group=group_id,
        )
```

File: Adapters/PerunRpcAdapter.py [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+                user["name"] = name
+
+                return User(**user)
+
+    def get_group_unique_name(
+        self,
+        attributes_api_instance: AttributesManagerApi,
+        group_name: str,
+        group_id: int,
+    ):
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:83]: You also need attribute name to get attribute.

**Review by xpavlic — CHANGES_REQUESTED** [review]

File: Adapters/PerunRpcAdapter.py [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+
+    def get_vo_attributes(
+        self, vo: VO, attr_names: List[str]
+    ) -> dict[str, Union[str, Optional[int], bool, List[str], dict[str, str]]]:
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            attr_names_map = AttributeUtils.get_rpc_attr_names(attr_names)
+            perun_attrs = attributes_api_instance.get_vo_attributes(
+                vo.id, attr_names_map.keys()
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:438]:
```suggestion
                vo.id, attrNames=attr_names_map.keys()
```

File: Adapters/PerunRpcAdapter.py [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+
+    def get_facility_attribute(
+        self, facility: Facility, attr_name: str
+    ) -> Union[str, Optional[int], bool, List[str], dict[str, str]]:
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            attr_name = AttributeUtils.get_rpc_attr_name(attr_name)
+            perun_attr = attributes_api_instance.get_attribute(
+                facility.id, attr_name
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:455]:
```suggestion
                facility=facility.id, attributeName=attr_name
```

File: config_templates/perun_rpc_config_template.yaml [hunk 11]
```diff
@@ -0,0 +1,3 @@
+"rpc_url": "my_url"
```
Comment (xpavlic) [inline config_templates/perun_rpc_config_template.yaml:1]: remove this file

**Review by xpavlic — CHANGES_REQUESTED** [review]

File: Adapters/PerunRpcAdapter.py (hunk 9 above)
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:438]: I've noticed that `get_vo_attributes` actually retrieves _all_ the attributes of given VO. I suggest we replace this call with calling `get_vo_attributes_by_names` which actually takes in a list of names to find.

File: Adapters/PerunRpcAdapter.py [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+
+    def get_user_attributes(
+        self, user: User, attr_names: List[str]
+    ) -> dict[str, Union[str, Optional[int], bool, List[str], dict[str, str]]]:
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            attr_names_map = AttributeUtils.get_rpc_attr_names(attr_names)
+            perun_attrs = attributes_api_instance.get_user_attributes(
+                user.id, attr_names_map.keys()
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:398]: Same case as in `get_vo_attributes`. I originally misunderstood the docs when writing this method. Now I'd replace it with `get_user_attributes_by_names`

File: Adapters/PerunRpcAdapter.py [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+            converted_groups = []
+            for group in member_groups:
+                group["unique_name"] = self.get_group_unique_name(
+                    attributes_api_instance, group["name"], group["id"]
+                )
+                converted_groups.append(Group(**group))
+
+            return converted_groups
+
+    def get_sp_groups(self, rp_identifier: str) -> List[Group]:
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:107]: I checked how this method is used in `simplesaml.php` and the return type is used as a `Group`. I've done a little experiment and changing the return type to `HasIdAbstract` here is going to silent the warning. However the warning will be propagated to subsequent calls of `Group` attributes other than `id` when working with the returned list of `Groups`. This problem was probably introduced when we decided to make `HasIdInterface` into `HasIdAbstract`. 

At this moment I'm not sure how to tackle this tricky type casting. I think we could drop the method `unique_entities` entirely and use `list(set(entities))` to filter out non-unique instances in a list of entities implementing custom `__hash__`  based on entity's `id`.

File: Adapters/PerunRpcAdapter.py [hunk 14]
```diff
@@ -0,0 +1,495 @@
...
+            return Facility(
+                perun_attr[0]["id"],
+                perun_attr[0]["name"],
+                perun_attr[0]["description"],
+                rp_identifier,
+            )
+
+    def get_users_groups_on_facility(
+        self, rp_identifier: str, user_id: str
+    ) -> List[Group]:
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:208]: The same thing as I mentioned in the above suggestion regarding `get_sp_groups`

File: Adapters/PerunRpcAdapter.py [hunk 15]
```diff
@@ -0,0 +1,495 @@
...
+
+    def get_facility_by_rp_identifier(
+        self,
+        rp_identifier: str,
+        rp_identifier_attr="perunFacilityAttr_entityID",
+    ) -> Facility:
+        with ApiClient(self.CONFIG) as api_client:
+            facilities_api_instance = FacilitiesManagerApi(api_client)
+
+            attr_name = AttributeUtils.get_rpc_attr_name(rp_identifier_attr)
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:180]: We currently have parameters `internal_attr_name` as well as `default_attr_name` in `AttributeUtils.get_rpc_attr_name`, however at the moment `rp_identifier_attr` is literally set to _"perunFacilityAttr_entityID"_ so this is the same as calling the method like this -> `get_rpc_attr_name(rp_identifier_attr, rp_identifier_attr)`.

Previously, we extracted the fallback **value** to `AttributeUtils` class, but the **key** to obtain that value is still _"perunFacilityAttr_entityID"_ in any case. I think we might want to get rid of the `default_attr_name` parameter because that's always the same according to my understanding. What we have stored in AttributeUtils.DEFAULT_ATTR_NAMES under that key is quivalent to _default attribute value_ tied to that `internal_attr_name`.

File: Adapters/PerunRpcAdapter.py [hunk 16]
```diff
@@ -0,0 +1,495 @@
...
+
+            return Facility(
+                perun_attr[0]["id"],
+                perun_attr[0]["name"],
+                perun_attr[0]["description"],
+                rp_identifier,
+            )
+
+    def get_users_groups_on_facility(
+        self, rp_identifier: str, user_id: str
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:207]: According to a convention we established when writing other methods I think we could _(should?)_ pass an entire `User` object as a parameter and access his/her `id` in the method body. I might be wrong though and will switch this to `int` for now.

File: Adapters/PerunRpcAdapter.py [hunk 17]
```diff
@@ -0,0 +1,495 @@
...
+
+    def get_facility_attributes(
+        self, facility: Facility, attr_names: List[str]
+    ) -> dict[str, Union[str, Optional[int], bool, List[str], dict[str, str]]]:
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            attr_names_map = AttributeUtils.get_rpc_attr_names(attr_names)
+            perun_attrs = attributes_api_instance.get_facility_attributes(
+                facility.id, attr_names_map.keys()
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:256]: I think I should have used the `get_facility_attributes_by_names` method here.

File: Adapters/PerunRpcAdapter.py [hunk 18]
```diff
@@ -0,0 +1,495 @@
...
+    def get_user_ext_source_attributes(
+        self, user_ext_source: UserExtSource, attr_names: List[str]
+    ) -> dict[str, Union[str, Optional[int], bool, List[str], dict[str, str]]]:
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            attr_names_map = AttributeUtils.get_rpc_attr_names(attr_names)
+            perun_attrs = (
+                attributes_api_instance.get_user_ext_source_attributes(
+                    user_ext_source.id, attr_names_map.keys()
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:286]: I think I should have used the `get_user_ext_source_attributes_by_names` method here.

File: Adapters/PerunRpcAdapter.py [hunk 19]
```diff
@@ -0,0 +1,495 @@
...
+            dict[
+                str, Union[str, Optional[int], bool, List[str], dict[str, str]]
+            ]
+        ],
+    ) -> None:
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            attributes_api_instance.set_user_ext_source_attributes(
+                user_ext_source.id, attributes
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:304]: I may be misunderstanding this but upon checking the `set_user_ext_source_attributes` method in `AttributesManagerApi` I can't see where the `user_ext_source` `id` would be extracted from the kwargs dict. I don't think attributes are the problem causing this warning, rather the `id` of ext. source which is unused in the generated method in my opinion.

File: Adapters/PerunRpcAdapter.py [hunk 20]
```diff
@@ -0,0 +1,495 @@
...
+
+            user_groups_ids = [user_group.id for user_group in user_groups]
+
+            for resource in resources:
+                resource_groups = resources_api_instance.get_assigned_groups(
+                    resource["id"]
+                )
+
+                resource_capabilities = attributes_api_instance.get_attribute(
+                    resource["id"],
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:360]: I agree this is not entirely correct right now but I think it should look like this:
```
resource_capabilities = attributes_api_instance.get_attribute(
                    attribute_id=resource["id"],
                    attribute_name="urn:perun:resource:attribute-def:def:capabilities"
                )
```

File: Adapters/PerunRpcAdapter.py [hunk 21]
```diff
@@ -0,0 +1,495 @@
...
+                    "urn:perun:resource:attribute-def:def:capabilities",
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:361]: Above reply concerns this comment as well.

File: Adapters/PerunRpcAdapter.py [hunk 22]
```diff
@@ -0,0 +1,495 @@
...
+        facility = self.get_facility_by_rp_identifier(rp_identifier)
+
+        if facility is None:
+            return []
+
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            facility_capabilities = attributes_api_instance.get_attribute(
+                facility.id,
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:384]: The same case as in resource above

File: Adapters/PerunRpcAdapter.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:360]: The second one should be right.

File: Adapters/PerunRpcAdapter.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:207]: We can pass User here.

File: Adapters/PerunRpcAdapter.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:256]: You are right.

File: Adapters/PerunRpcAdapter.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:286]: You are right.

File: Adapters/PerunRpcAdapter.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:398]: Right.

File: Adapters/PerunRpcAdapter.py (hunk 9 above)
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:438]: Right

File: Adapters/PerunRpcAdapter.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:361]: Same pattern as above.

File: Adapters/PerunRpcAdapter.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:107]: I figured out, that warning is caused by wrong imports in Model classes, which I already suggested you to fix.

**Review by xpavlic — CHANGES_REQUESTED** [review]
Added comments.

File: Adapters/PerunRpcAdapter.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:208]: I looked into code once more and in old implementation, you could send attribute name into function (does not have to be perunFacilityAttr_entityID) and if it is not found, then still use default value. I think, we can remove rp_identifier_attr from parameter and make it fixed value. Then we can remove default_attr_name and change AttributeUtils.

File: Adapters/PerunRpcAdapter.py [hunk 23]
```diff
@@ -0,0 +1,531 @@
...
+
+        if facility is None:
+            return []
+
+        with ApiClient(self.CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            facility_capabilities = attributes_api_instance.get_attribute(
+                facility=facility.id,
+                attribute_name="urn:perun:facility:attribute-def:def"
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:419]: I think this should be also read from config by AttributeUtils

File: Adapters/PerunRpcAdapter.py [hunk 24]
```diff
@@ -0,0 +1,531 @@
...
+        # TODO replace config_data parameter by access to a static object
+        #  with parsed configs, possibly remove try-except block if
+        #  config_data is already validated by said object.
+
+        # TODO AttributeUtils is used in this class but it's no longer
+        #  static and requires its own config as well. Make sure it's
+        #  instantiated here or passed to constructor.
+
+        self.logger = Logger.get_logger(self.__class__.__name__)
+        self.CONFIG_FILE_PATH = "config_templates/attribute_map.yaml"
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:35]: I think this is not needed.

File: Adapters/PerunRpcAdapter.py [hunk 25]
```diff
@@ -0,0 +1,531 @@
...
+                self.CONFIG.access_token = config_data["access_token"]
+            elif auth_type == self.API_KEY_AUTH:
+                self.CONFIG.api_key[self.API_KEY_AUTH] = config_data["api_key"]
+            else:
+                self.logger.warning(
+                    f'Authentication type "{auth_type}" is not a supported '
+                    f'way of authentication, please set "auth_type" to one '
+                    f'of "{self.BASIC_AUTH}", "{self.BEARER_AUTH}" or "'
+                    f'{self.API_KEY_AUTH}"'
+                )
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:58]: I think this should cause exception, because we can not instantiate PerunRpcAdapter

File: Adapters/PerunRpcAdapter.py [hunk 26]
```diff
@@ -0,0 +1,531 @@
...
+
+        except KeyError as error:
+            self.logger.warning(
+                f"{error} key not found in config file located at "
+                f"{self.CONFIG_FILE_PATH}"
```
Comment (xpavlic) [inline Adapters/PerunRpcAdapter.py:63]: We dont need filepath, just write in config instead. And We should rethrow exception.

**Review by xpavlic — CHANGES_REQUESTED** [review]

File: Adapters/PerunRpcAdapter.py (hunk 25 above)
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:58]: Should we keep this log and raise, say, `ValueError` on top of that or remove the logging entirely and instead pass its error message in the exception?

File: Adapters/PerunRpcAdapter.py (hunk 26 above)
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:63]: Similar as above, should we move the entire message to the new exception or keep the logging statement as well?

File: Adapters/PerunRpcAdapter.py [hunk 27]
```diff
@@ -0,0 +1,531 @@
...
+
+    def get_group_unique_name(
+            self,
+            attributes_api_instance: AttributesManagerApi,
+            group_name: str,
+            group_id: int,
+    ) -> str:
+        attr = attributes_api_instance.get_attribute(
+            group=group_id,
+            attribute_name="urn:perun:group:attribute-def:virt:voShortName",
```
[the remark this answers is not in the document]
Comment (PeterBolha) [reply Adapters/PerunRpcAdapter.py:96]: I just noticed that this `attribute_name` is taken from `AdapterRpc` where it's used in the method `getMemberGroups` like this. However there's only `urn:perun:group:attribute-def:virt:voShortNames` defined in the config file. I'm not sure if this is a typo, but our config would return a method with return type `list` and `attribute_name` should be a single value.

[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:96]: It is not a typo it is just value which is not in config. Just add it there.

File: Adapters/PerunRpcAdapter.py (hunk 25 above)
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py:58]: I think propagation of exception should be enough. Also we will need manage exceptions from adapters in AdaptersManager, to be able to switch between adapters.

File: Adapters/PerunRpcAdapter.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (xpavlic) [reply Adapters/PerunRpcAdapter.py]: I found out that in RPC you need to make group unique name from voShortName and group name, but in LDAP there is just attribute groupUniqueName. So voShortName doesnt have its equivalent in ldap and I think this is the reason, that attribute name is defined here. After this finding, I would let attribute name same as i was. (same for other cases).

File: adapters/PerunRpcAdapter.py [hunk 28]
```diff
@@ -0,0 +1,565 @@
...
+            group_name: str,
+            group_id: int,
+    ) -> str:
+        attr = attributes_api_instance.get_attribute(
+            group=group_id,
+            attribute_name="urn:perun:group:attribute-def:virt:voShortName",
+        )
+        return f'{attr["value"]}:{group_name}'
+
+    def _create_internal_representation_groups(self,
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:103]: I think, that right type of input_groups is List[perun_openapi.model.group.Group] (model from openApi). Also add import perun_openapi.model.group to imports.

File: adapters/PerunRpcAdapter.py [hunk 29]
```diff
@@ -0,0 +1,565 @@
...
+                perun_attr[0]["name"],
+                perun_attr[0]["description"],
+                rp_identifier,
+            )
+
+    def get_users_groups_on_facility(
+            self, rp_identifier: str, user: User
+    ) -> List[Group]:
+        facility = self.get_facility_by_rp_identifier(rp_identifier)
+        groups = []
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:253]:
```suggestion
```

File: adapters/PerunRpcAdapter.py [hunk 30]
```diff
@@ -0,0 +1,565 @@
...
+
+        if facility is None:
+            return groups
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:256]:
```suggestion
            return []
```

File: adapters/PerunRpcAdapter.py [hunk 31]
```diff
@@ -0,0 +1,565 @@
...
+            facility = self.get_facility_by_rp_identifier(rp_identifier)
+            if facility is None:
+                return []
+            perun_attrs = (
+                facilities_api_instance.get_assigned_resources_for_facility(
+                    facility.id
+                )
+            )
+
+            resources_ids = [perun_attr.id for perun_attr in perun_attrs]
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:167]:
```suggestion
            resources = (
                facilities_api_instance.get_assigned_resources_for_facility(
                    facility.id
                )
            )

            resources_ids = [resource.id for resource in resources]
```

File: adapters/PerunRpcAdapter.py [hunk 32]
```diff
@@ -0,0 +1,565 @@
...
+            rp_identifier: str,
+    ) -> Optional[Facility]:
+        with ApiClient(self._CONFIG) as api_client:
+            facilities_api_instance = FacilitiesManagerApi(api_client)
+
+            attr_name = self._ATTRIBUTE_UTILS.get_rpc_attr_name(
+                self._RP_ID_ATTR
+            )
+
+            perun_attr = facilities_api_instance.get_facilities_by_attribute(
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:226]: Rename perun_attr to facilities.

File: adapters/PerunRpcAdapter.py [hunk 33]
```diff
@@ -0,0 +1,565 @@
...
+                users_groups_on_facility, converted_groups,
+                attributes_api_instance)
+            return converted_groups
+
+    def _get_rp_id(self, facility: Facility) -> str:
+        return self.get_facility_attributes(facility, [self._RP_ID_ATTR]) \
+            .get(self._RP_ID_ATTR)
+
+    # TODO add support for searching by more attributes to OpenAPI
+    #  (implement searcher in FacilityManager)
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:279]:
```suggestion
    #  (reimplement with Searcher, not working now)
```

File: adapters/PerunRpcAdapter.py [hunk 34]
```diff
@@ -0,0 +1,565 @@
...
+            )
+            user_groups_ids = [user_group.id for user_group in user_groups]
+            for resource in resources:
+                resource_groups = resources_api_instance.get_assigned_groups(
+                    resource["id"]
+                )
+
+                resource_capabilities = attributes_api_instance.get_attribute(
+                    resource=resource["id"],
+                    attribute_name="urn:perun:resource:attribute-def:def"
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:427]: get attribute_name from AttributeUtils (from config)

File: adapters/PerunRpcAdapter.py [hunk 35]
```diff
@@ -0,0 +1,565 @@
...
+
+        if facility is None:
+            return []
+
+        with ApiClient(self._CONFIG) as api_client:
+            attributes_api_instance = AttributesManagerApi(api_client)
+
+            facility_capabilities = attributes_api_instance.get_attribute(
+                facility=facility.id,
+                attribute_name="urn:perun:facility:attribute-def:def"
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:451]: same as above

File: adapters/PerunRpcAdapter.py [hunk 36]
```diff
@@ -0,0 +1,565 @@
...
+            )["value"]
+
+            return facility_capabilities
+
+    def get_user_attributes(
+            self, user: User, attr_names: List[str]
+    ) -> dict[str, Union[str, Optional[int], bool, List[str], dict[str, str]]]:
+        default_attribute_name = "perunUserAttribute_loa"
+        if not attr_names:
+            attr_names.append(default_attribute_name)
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:462]: I am not sure, why you are adding loa to default attribute names.

**Review by xpavlic — CHANGES_REQUESTED** [review]
Please also remove changes in openApi. As it is automatically generated, we should not intervene manually

This should be last things to change.

File: adapters/PerunRpcAdapter.py (hunk 36 above)
Comment (PeterBolha) [reply adapters/PerunRpcAdapter.py:462]: This was a special case we discussed at one point. OpenApi's `AttributesManagerApi.get_user_attributes_by_names` returns `loa` attribute if the input is an empty array (if we don't ask for any attributes, it returns `loa` nontheless). Our implementation then pairs input attributes as keys with obtained attribute values but it arrives at an error if we recieve a value for a key we didn't ask for, therefore it isn't in the input array.

Simply put, we recieve value for an input we didn't ask for. This assignment in case of an empty input makes it explicitly ask for the default value so this implementation behaves as OpenAPI and doesn't crash with KeyError. It seems quite strange to me so perhaps I'm misunderstanding what's going on.

The same thing is happening in `get_vo_attributes`

File: adapters/PerunRpcAdapter.py (hunk 35 above)
Comment (PeterBolha) [reply adapters/PerunRpcAdapter.py:451]: We also discussed this on Slack if I remember correctly. We tried to extract these names to the config file but upon further inspection, we discovered that these attrs were only used by RPC adapter so we decided to revert the extraction and place them back to the rpc adapter.

File: adapters/PerunRpcAdapter.py [hunk 37]
```diff
@@ -0,0 +1,634 @@
...
+            self._create_internal_representation_groups(
+                users_groups_on_facility, converted_groups,
+                attributes_api_instance)
+            return converted_groups
+
+    def _get_rp_id(self, facility: Facility) -> str:
+        return self.get_facility_attributes(facility, [self._RP_ID_ATTR]) \
+            .get(self._RP_ID_ATTR)
+
+    # TODO add support for searching by more attributes to OpenAPI
```
Comment (xpavlic) [inline adapters/PerunRpcAdapter.py:293]: Use Searcher instead as it is already implemented in openApi.

**Review by xpavlic — CHANGES_REQUESTED** [review]

**Review by xpavlic — CHANGES_REQUESTED** [review]
Overall looks good. Squash commits please.

**Review by xpavlic — APPROVED** [review]
